experimental/tools/handshake-simulator/ExecModels.cpp

```diff
@@ -162,207 +261,541 @@ LogicalResult dynamatic::experimental::initialiseMap(
 namespace dynamatic {
 namespace experimental {
 
-/// Default CIRCT fork
-bool DefaultFork::tryExecute(llvm::DenseMap<Value, llvm::Any> &valueMap,
-                             llvm::DenseMap<unsigned, unsigned> &memoryMap,
-                             llvm::DenseMap<Value, double> &timeMap,
-                             std::vector<std::vector<llvm::Any>> &store,
-                             SmallVector<Value> &scheduleList, ModelMap &models,
-                             circt::Operation &opArg) {
+//--- Default CIRCT models ---------------------------------------------------//
 
+bool DefaultFork::tryExecute(ExecutableData &data, circt::Operation &opArg) {
   auto op = dyn_cast<circt::handshake::ForkOp>(opArg);
   auto executeFunc = [](std::vector<llvm::Any> &ins,
                         std::vector<llvm::Any> &outs, circt::Operation &op) {
     for (auto &out : outs)
       out = ins[0];
   };
-  return tryToExecute(op.getOperation(), valueMap, timeMap, scheduleList,
-                      models, executeFunc, 1);
+  return tryToExecute(op.getOperation(), data.valueMap, data.models,
+                      executeFunc);
 }
 
-/// Default CIRCT merge
-bool DefaultMerge::tryExecute(llvm::DenseMap<Value, llvm::Any> &valueMap,
-                              llvm::DenseMap<unsigned, unsigned> &memoryMap,
-                              llvm::DenseMap<Value, double> &timeMap,
-                              std::vector<std::vector<llvm::Any>> &store,
-                              SmallVector<Value> &scheduleList,
-                              ModelMap &models, circt::Operation &opArg) {
+bool DefaultMerge::tryExecute(ExecutableData &data, circt::Operation &opArg) {
   auto op = dyn_cast<circt::handshake::MergeOp>(opArg);
   bool found = false;
   for (Value in : op.getOperands()) {
-    if (valueMap.count(in) == 1) {
+    if (data.valueMap.count(in) == 1) {
```

Comment:
Checking this for the first time, but it's actually fine if there is more than one valid input. We say the merge is "undeterministic" in the sense that in such situations it will just forward one of its valid inputs. In practice of course it's deterministic (circuits always are). So the model should just stop at the first valid value it encounters. The same is true for the control merge.

If this was intended for the next PR then just keep this in mind, otherwise fix now.